src/NServiceBus.Core/Routing/EndpointInstance.cs

```diff
@@ -56,6 +56,24 @@ public EndpointInstance SetProperty(string key, string value)
         }
 
         /// <summary>
+        /// Creates a new endpoint instance with specified discriminator.
+        /// </summary>
+        /// <param name="discriminator">New discriminator value.</param>
+        public EndpointInstance SetDiscriminator(string discriminator)
```

Comment:
Alternatively, is there a reason these need to return new instances vs. updating the existing instance with the new values?